### DISCUSSION TITLE
Why every method from `Location` returns a `Result<T, JsValue>`?

### DISCUSSION THREAD

**Discussion by xeho91** [wasm-bindgen/wasm-bindgen#3713] [Q&A] [answered]
Hi `web-sys` maintainers!

I'd like to understand the reasoning behind the setting return value of every methods from the [`Location`](https://docs.rs/web-sys/latest/web_sys/struct.Location.html#) with `Result<T, JsValue>`.

---

Lets take [`Location::origin(&self)`](https://docs.rs/web-sys/latest/web_sys/struct.Location.html#method.origin) for example.

**Following the [MDN WebAPI for `window.location.origin`](https://developer.mozilla.org/en-US/docs/Web/API/Location/origin) it never returns an error.**

Why then, instead of returning `String` it is wrapped with `Result`, and returning `JsValue` when it fails?
What is `JsValue` supposed to return exactly?

Trying to study and understand by myself, I found out that `window().unwrap().location()` doesn't return a `Result`, which would give me a suggestion that this is related to some conversion mechanism from JavaScript to Rust. But at this point, I am clueless.

**Comment by Liamolucko** [comment, marked as answer]
`web-sys` is automatically generated from a pile of 'WebIDL' files [here](https://github.com/rustwasm/wasm-bindgen/tree/main/crates/web-sys/webidls), which is how web APIs define their interfaces. In particular, it decides whether or not a function or getter should return a `Result` based on whether it's annotated with a `[Throws]` attribute in WebIDL, which `location.origin` is:

https://github.com/rustwasm/wasm-bindgen/blob/bb12058e1facd26c5c647282a125088535be577e/crates/web-sys/webidls/enabled/Location.webidl#L25-L26

...Except, the `[Throws]` attribute isn't actually an official part of WebIDL. It's a custom addition used by Firefox, which is where `web-sys` copied most of its WebIDL from. So, the answer to why it returns `Result` is pretty much 'because Firefox thought it should back when we copied its WebIDL'.

(Copying Firefox's WebIDL instead of using the versions from the actual standards has led to some issues before, e.g. #3227, although I can't remember any particular cases where something was incorrectly annotated with `[Throws]`. It might have happened though.)

As for whether or not it _should_ return `Result` though, I think the answer is yes. My rule of thumb for deciding whether or not a function should be annotated with `[Throws]` when adding a new API is to check the specification for that function and see if there are any cases where it says to throw an error, since that should be the definitive source. And [the spec for `location.origin`](https://html.spec.whatwg.org/multipage/nav-history-apis.html#dom-location-origin) says:

> The origin getter steps are:
>
> 1. If [this](https://webidl.spec.whatwg.org/#this)'s [relevant Document](https://html.spec.whatwg.org/multipage/nav-history-apis.html#relevant-document) is non-null and its [origin](https://dom.spec.whatwg.org/#concept-document-origin) is not [same origin-domain](https://html.spec.whatwg.org/multipage/browsers.html#same-origin-domain) with the [entry settings object](https://html.spec.whatwg.org/multipage/webappapis.html#entry-settings-object)'s [origin](https://html.spec.whatwg.org/multipage/webappapis.html#concept-settings-object-origin), then throw a ["SecurityError"](https://webidl.spec.whatwg.org/#securityerror) [DOMException](https://webidl.spec.whatwg.org/#dfn-DOMException).
>
> 2. Return the [serialization](https://html.spec.whatwg.org/multipage/browsers.html#ascii-serialisation-of-an-origin) of [this](https://webidl.spec.whatwg.org/#this)'s [url](https://html.spec.whatwg.org/multipage/nav-history-apis.html#concept-location-url)'s [origin](https://url.spec.whatwg.org/#concept-url-origin).

Apparently it can throw a `SecurityError` in some scenarios, and so `Location::origin` should return a `Result` so that it can be properly handled.

**Reply by xeho91** [reply]
Thanks for such detailed answer, have a good time!
